### PR TITLE
Raise meaningful error when permission denied to copy JT

### PR DESCRIPTION
**Pull request by ghost**
##### SUMMARY

When a user doesn't have access to all the credentials of a job template, they cannot copy the JT. However, currently we raise a default `PermissionDenied`, which doesn't give the user insight into what's wrong. So, this PR just adds a custom message indicating that access to credentials is missing.

Signed-off-by: Vismay Golwala <vgolwala@redhat.com>

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

 - API

##### AWX VERSION

```
awx: 4.0.0
```

### CODE REVIEW

**Review by matburt — DISMISSED** [review]
I think this looks good if we can add one more test that involves the user *having* access to the credential and being able to copy.

**Review by AlanCoding — CHANGES_REQUESTED** [review]
This is too much slowdown in the templates list, which is one of the most highly trafficked things. Instead, I would prefer a solution that the JobTemplateAccess method `can_copy_related` does `raise PermissionDenied` with a message that the user does not have permission to the related credentials. That way, we don't check it in the list view, and the user has guidance about how to resolve the situation.

**Comment by ghost** [comment]
> This is too much slowdown in the templates list, which is one of the most highly trafficked things. Instead, I would prefer a solution that the JobTemplateAccess method `can_copy_related` does `raise PermissionDenied` with a message that the user does not have permission to the related credentials. That way, we don't check it in the list view, and the user has guidance about how to resolve the situation.

@AlanCoding I think what you are suggesting is a way to validate `copy` request. However, that's already been handled. So even without this PR, if we try to copy a JT, it will raise `PermissionDenied` error. But, if a user isn't allowed to do so, s/he should not be shown an icon to copy JT. That's what I am trying to achieve with this. So, this fix updates the `user_capabilities`, such that, copy icon gets hidden in the UI. It is in reference to Tower issue 3491. Hope it makes sense.

**Comment by AlanCoding** [comment]
What I'm suggesting is an alternative which shouldn't require structural changes to how the copy checks happen. The reported issue is that the "copy" capability returns True in this case, and you're flipping it to False, but the cost for doing this is very high in terms of performance and maintainability. Raising an exception with just a few lines of text [here](https://github.com/ansible/awx/blob/4bfab304b065c714a59a6949c1e76aca6838453e/awx/main/access.py#L1501) should be able to address the complaint, and will avoid changing anything about what gets called where.

If you followed your approach here to its logical conclusion, I think you would find that `can_copy_related` isn't needed at all. Its logic would be much more logically absorbed into `can_copy`. But the introduction of `user_capabilities` was based on the assumption that we could get performant data directly from the role ancestors. In cases where we can't, custom error messages are pretty good, and IMHO, better for the user experience.

**Comment by AlanCoding** [comment]
Instead of raising in the generic view (where it is now), can we raise the exception at this point, in the event that the determination is False?

https://github.com/ansible/awx/blob/46206db659363546357122c740439a7016f33a50/awx/main/access.py#L1510

I wonder if you tried that and had some unexpected problems with it?

**Comment by ghost** [comment]
Done fixing API error

@AlanCoding does it look accurate now? 

I tried it before and it worked, but I thought raising in the generic view was a smaller change. But then, it was in the wrong place and I updated it now.

**Comment by AlanCoding** [comment]
That looks good to me.

I think you should mark it for translation, because it's user-facing with the `_()` syntax used in other places in that file.

I see a test fail, but that probably need to be addressed in the test

File: awx/main/tests/functional/test_copy.py [hunk 1]
```diff
@@ -20,10 +20,10 @@ def test_job_template_copy(post, get, project, inventory, machine_credential, va
     job_template_with_survey_passwords.admin_role.members.add(alice)
     project.admin_role.members.add(alice)
     inventory.admin_role.members.add(alice)
-    assert get(
+    get(
         reverse('api:job_template_copy', kwargs={'pk': job_template_with_survey_passwords.pk}),
-        alice, expect=200
-    ).data['can_copy'] is False
+        alice, expect=403
```
Comment (AlanCoding) [inline awx/main/tests/functional/test_copy.py]: oh, I think I see what happened. For the get request, a 200 would be preferable for alice, who has admin for the job template.

I think we just want to catch `PermmissionDenied` in

https://github.com/ansible/awx/blob/cf2ad1486a293906479961696b623086bccc53f7/awx/api/generics.py#L963

And give `False` for the value in that case.

**Review by AlanCoding — APPROVED** [review]
Excellent, thanks for all the work you put in, this looks great

**Comment by ryanpetrello** [comment]
regate

**Pull request merged**